### PR TITLE
Fix #5044: guard against forward references in the TypeTree of New trees

### LINKED ISSUE #5044: NoSuchElementException in the backend.
```scala
class I0 {
  class I1
  def i3 = {
    val i4 = new i5.I1
    val i5 = new I0
  }
}
```
```scala
Exception in thread "main" java.util.NoSuchElementException: val i5
	at scala.collection.mutable.AnyRefMap$ExceptionDefault.apply(AnyRefMap.scala:425)
	at scala.collection.mutable.AnyRefMap$ExceptionDefault.apply(AnyRefMap.scala:424)
	at scala.collection.mutable.AnyRefMap.apply(AnyRefMap.scala:180)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder$locals$.load(BCodeSkelBuilder.scala:401)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:391)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadArguments$$anonfun$1(BCodeBodyBuilder.scala:1018)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:18)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadArguments(BCodeBodyBuilder.scala:1018)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genApply(BCodeBodyBuilder.scala:683)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:337)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:290)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genStat(BCodeBodyBuilder.scala:89)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genNormalBlock$2$$anonfun$1(BCodeBodyBuilder.scala:860)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:18)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genNormalBlock$1(BCodeBodyBuilder.scala:860)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genBlock(BCodeBodyBuilder.scala:892)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:408)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.emitNormalMethodBody$1(BCodeSkelBuilder.scala:602)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genDefDef(BCodeSkelBuilder.scala:635)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:506)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen$$anonfun$1(BCodeSkelBuilder.scala:508)
```
NSEE-b300a87b772c130ad27152cdbb359c7767348152

### PR DESCRIPTION
**Pull request by Jasper-M**
The extra dealiasing step also avoids [this scalac bug](https://github.com/scala/bug/issues/11177).

### CODE REVIEW

**Comment by Jasper-M** [comment]
Should the following compile?

```
scala> def foo = { type T = Foo; val a = new T; class Foo }
def foo: Unit
```

Scalac doesn't think so.

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 1]
```diff
@@ -1010,9 +1010,14 @@ class RefChecks extends MiniPhase { thisPhase =>
   }
 
   override def transformNew(tree: New)(implicit ctx: Context) = {
-    val sym = tree.tpe.typeSymbol
+    val tpe = tree.tpe
+    val sym = tpe.typeSymbol
     checkUndesiredProperties(sym, tree.pos)
     currentLevel.enterReference(sym, tree.pos)
+    tpe.dealias.foreachPart {
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: Is `dealias` required?

File: tests/neg/i5044b.scala [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class I0 {
```
Comment (allanrenucci) [inline tests/neg/i5044b.scala]: I would merge the two tests together:
```scala
class I0 {
  class I1
  def test0 = {
    val x = new y.I1 // error: `y` is a forward reference extending over the definition of `x`
    val y = new I0
  }

  def test1 = {
    type T = y.I1
    val x = new T // error: `y` is a forward reference extending over the definition of `x`
    val y = new I0
  }
}
```

File: tests/neg/i5044b.scala [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+  class I1
+  def i3 = {
+    type T = i5.I1
+    val i4 = new T // error: `i5` is a forward reference extending over the definition of `i4`
+    val i5 = new I0
+  }
```
Comment (allanrenucci) [inline tests/neg/i5044b.scala]: Here is another test with a type lambda
```scala
class I0 {
  class I2[T1, T2]
  def test2 = {
    type A[T] = i5.I2[T, String]
    val i4 = new A[Int]
    val i5 = new I0
  }
}
```
This one currently crashes with Scala 2.12.6 😄

Comment (allanrenucci) [reply tests/neg/i5044b.scala]: I would also add:
```scala
  def test3 = {
    val x = new T // error: `T` is a forward reference extending over the definition of `x`
    val y = new I0
    type T = y.I1
  }
```

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala (hunk 1 above)
Comment (Jasper-M) [reply compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: Without it the second test still fails. If you would dealias a bit sooner I think `def foo = { type T = Foo; val a = new T; class Foo }` would also emit an error, like in scalac.

**Pull request merged**